### PR TITLE
Expose `sql_to_statement` and `statement_to_plan` on `SessionState`

### LINKED ISSUE #4957: Split `create_logical_plan()` into parsing and planning
**Describe the bug**

It's not really a bug, more of an API regression. When #4607 got merged, consumers of the DataFusion public API lost the ability to (easily) inspect or alter the AST of a SQL statement before it got planned. This is because the function that runs it became non-trivial in that it copies tables into a new context, so now the consumer would either have to duplicate the table copying logic, or go without.

**To Reproduce**

This used to work:

```
        let statements = DFParser::parse_sql(sql)?;
        let statements: Vec<_> = statements.into_iter().collect();

       // Do something with AST here

        let state = df_ctx.state();
        let planner = SqlToRel::new(&state);
        let plan = planner.statement_to_plan(stmt).map_err(|e| err!("", e))?;
        let df = Arc::new(DataFrame::new(df_ctx.state(), plan));
```

But now it yields: `ContextProvider not implented for SessionState`

**Expected behavior**

A way to parse, manipulate AST, then plan separately.

### PR DESCRIPTION
**Pull request by avantgardnerio**
# Which issue does this PR close?

Closes #4957.

# Rationale for this change

Described in issue.

# What changes are included in this PR?

Splitting planning and parsing.

# Are these changes tested?

Transitively, yes.

# Are there any user-facing changes?

New API methods, no breaking changes.

### CODE REVIEW

File: datafusion/core/src/execution/context.rs [hunk 1]
```diff
@@ -1636,22 +1636,34 @@ impl SessionState {
...
+
+    /// Convert an ast Statement into a LogicalPlan
+    pub async fn statement_to_plan(
+        &self,
+        statement: datafusion_sql::parser::Statement,
+    ) -> Result<LogicalPlan> {
+        use crate::catalog::information_schema::INFORMATION_SCHEMA_TABLES;
+        use datafusion_sql::parser::Statement as DFStatement;
+        use sqlparser::ast::*;
+        use std::collections::hash_map::Entry;
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1666]: These `use` statements seem out of place 🤔

**Review by alamb — APPROVED** [review]
I think this API makes a lot of sense. 

cc @tustvold  who I think had some thoughts in this area, but I believe this PR keeps the same spirit of encapsulation.

**Review by tustvold — APPROVED** [review]
Looks good to me :+1:

File: datafusion/core/src/execution/context.rs [hunk 2]
```diff
@@ -1729,6 +1741,15 @@ impl SessionState {
...
     }
 
+    /// Creates a [`LogicalPlan`] from the provided SQL string
+    ///
+    /// See [`SessionContext::sql`] for a higher-level interface that also handles DDL
+    pub async fn create_logical_plan(&self, sql: &str) -> Result<LogicalPlan> {
+        let statement = self.sql_to_statement(sql)?;
+        let plan = self.statement_to_plan(statement).await?;
+        Ok(plan)
+    }
```
[the remark this answers is not in the document]
Comment (avantgardnerio) [reply datafusion/core/src/execution/context.rs:1751]: @gruuya your suggestion makes sense and I'd be amenable to merging a PR containing the changes you described, but in the interest of expedience, given the approvals on this PR, I'd like to merge it as-is.

If you file an issue, I can take a look next time I'm free.

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/execution/context.rs:1751]: > Namely, while this solution handles the stated case well it doesn't address the general problem of SqlToRel now being hard/impossible to instantiate properly (with the new trait bound for ContextProvider, which is only implemented for a private struct, and the entire information_schema module also being private further complicating a custom implementation of ContextProvider). This in turn forbids usage of other useful public SqlToRel methods, such as sql_to_expr.

@gruuya I don't think we have contemplated using `SqlToRel` directly in the past. If that is a usecase that is important, what I found works well is to make an example demonstrating what you are doing. Here are some examples:

* https://github.com/apache/arrow-datafusion/blob/master/datafusion-examples/examples/rewrite_expr.rs
* https://github.com/apache/arrow-datafusion/blob/master/datafusion-examples/examples/custom_datasource.rs


An example has several advantages:
1. It ensures your usecase is clearly documented so people who are changing DataFusion know it is used and don't break it accidentally 
3. It then gives you an example to start from if the interface changes (you can look at the example to see how the interface may have changed)
4. It helps others who might have the same need

[the remark this answers is not in the document]
Comment (gruuya) [reply datafusion/core/src/execution/context.rs:1751]: That's a fair point; let me see if I can make a self-contained example for the usecase we have and then try my luck with a PR. 

Namely, we use `sql_to_expr` to [convert](https://github.com/splitgraph/seafowl/blob/main/src/context.rs#L1173) the `WHERE` clauses in `UPDATE` and `DELETE` statements from sqlparser `Expr`s into Datafusion `Expr`s during logical planning. Subsequently, during physical planning of our custom logical extension nodes for `UPDATE` and `DELETE` we employ these expressions in order to achieve the desired action using available Datafusion primitives (pruning, scanning, filtering, etc.).

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/execution/context.rs:1751]: In parallel you may be interested in some work @avantgardnerio  is working on here https://github.com/apache/arrow-datafusion/pull/4902 and likely elsewhere to add better INSERT/UPDATE/DELETE support directly to DataFusion

[the remark this answers is not in the document]
Comment (gruuya) [reply datafusion/core/src/execution/context.rs:1751]: Oh, interesting; will check it out thanks!

**Pull request merged**